azurelinuxagent/common/protocol/wire.py

```diff
@@ -913,49 +913,51 @@ def _call_hostplugin_with_container_check(self, host_func):
+        return ret
 
-        if not HostPluginProtocol.is_default_channel():
-            ret = None
-            try:
-                ret = direct_func()
+    def __send_request_using_direct_channel(self, direct_func):
+        ret = None
+        try:
+            ret = direct_func()
 
-                # Different direct channel functions report failure in different ways: by returning None, False,
-                # or raising ResourceGone or InvalidContainer exceptions.
-                if not ret:
-                    logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                            "switching to host plugin.")
-            except (ResourceGoneError, InvalidContainerError) as e: # pylint: disable=C0103
+            # Different direct channel functions report failure in different ways: by returning None, False,
+            # or raising ResourceGone or InvalidContainer exceptions.
+            if not ret:
                 logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                        "switching to host plugin. Error: {0}".format(ustr(e)))
+                                                        "trying host plugin next.")
+        except (ResourceGoneError, InvalidContainerError) as e:  # pylint: disable=C0103
+            logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
+                                                    "trying host plugin next. Error: {0}".format(ustr(e)))
 
-            if ret:
-                return ret
-        else:
-            logger.periodic_info(logger.EVERY_HALF_DAY, "[PERIODIC] Using host plugin as default channel.")
+        if ret and HostPluginProtocol.is_default_channel():
+            # Set direct channel as default if it succeeded.
+            logger.info("Switching default channel to WireServer.")
+            HostPluginProtocol.set_default_channel(False)
 
-        ret = self._call_hostplugin_with_container_check(host_func)
+        return ret
 
-        if not HostPluginProtocol.is_default_channel():
-            logger.info("Setting host plugin as default channel from now on. "
-                        "Restart the agent to reset the default channel.")
-            HostPluginProtocol.set_default_channel(True)
+    def send_request_using_appropriate_channel(self, direct_func, host_func):
+        primary_ret = None
+        try:
+            primary_ret = self.__send_request_using_host_channel(host_func) if HostPluginProtocol.is_default_channel() \
+                else self.__send_request_using_direct_channel(direct_func)
+        except Exception:
+            pass
 
-        return ret
+        if primary_ret:
+            return primary_ret
+
+        secondary_ret = self.__send_request_using_direct_channel(direct_func) if HostPluginProtocol.is_default_channel() \
```

Comment:
we need to improve the clarity of this code... it is really confusing that we are using the direct channel if the default channel is the host ga plugin

this condition is relying in non-obvious side-effects of the previous calls

Reply:
> it is really confusing that we are using the direct channel if the default channel is the host ga plugin

hence the name "secondary" :smile:

Reply:
huh?

Reply:
what i was trying to ask is that when HostPluginProtocol.is_default_channel() == True is supposed to indicate that we should use the host channel and this is using the direct channel

Reply:
As discussed offline, I'll rewrite this method to be clearer.